quint/src/parsing/ToIrListener.ts

```diff
@@ -1270,13 +1241,39 @@ export class ToIrListener implements QuintListener {
   // pop a row
   private popRow(): Row {
     // the user has specified a row
-    const tp = this.rowStack.pop()
-    return tp!
+    return this.rowStack.pop() ?? { kind: 'empty' }
   }
 
-  private undefinedDef(ctx: any): QuintEx {
-    const id = this.getId(ctx)
-    return { id, kind: 'bool', value: true }
+  private undefinedExpr(ctx: any): () => QuintEx {
+    return () => {
+      const id = this.getId(ctx)
+      return { id, kind: 'bool', value: true }
+    }
+  }
```

Comment:
Should we introduce an `UndefinedExpr` node in the IR so we can be very clear about where holes are showing up? I'm worried we could end up with weird bugs where parts of the AST are replaced with `true` and it will be hard for maintainers to find out why that would be happening.

Reply:
Hmm that might be a good idea for the next step. For now, the IR with undefined components should only be manipulated in name resolution. After that, we halt the process and report all errors gathered so far. This is not something guaranteed, but I'm pretty confident that we are ensuring that this is the case.

For the next steps, I want the type checker and effect checker to run over this IR as well, and my original plan was to have a check like: if there is an error for the id of this component, don't try to infer type/effect for the component.

Perhaps introducing undefined components to the IR, as you suggest, is a better alternative than relying on the errors. This would be a bigger change tho, as it requires another case in many IR-related functions and tests.

Reply:
I see. Maybe something to consider for a followup, then.

Reply:
I like that idea!